Fix PaliGemma Pad Token Masking During Training #35855

sambhavnoobcoder opened the pull request:
## Problem Statement
In PaliGemma model's `_update_causal_mask` function, padding tokens were being incorrectly unmasked during training mode. This occurred because the order of operations first applied padding masks and then unmasked prefix tokens (including pad tokens) during training, leading to inconsistent behavior especially with left padding.

Fixes : #35855 

## Approach
After analyzing the masking logic, we identified that the issue stemmed from the sequence of operations in mask application. The solution was to reorder the masking operations to:
1. First unmask prefix tokens during training mode
2. Then apply padding masks

This ensures pad tokens remain masked regardless of their position or training state.

## Implementation
The fix involved reordering the masking operations in the `_update_causal_mask` function while maintaining the same mathematical operations. This approach ensures:
- No changes to the underlying logic
- Minimal code changes
- Preserved backward compatibility
- Consistent behavior across training and inference modes

### Test Coverage
The new test validates:
1. Pad tokens remain masked (value = dtype.min) during training
2. Non-pad tokens are properly masked/unmasked based on their position
3. Behavior is consistent across different batch sizes and sequence lengths

### Screenshots

<img width="617" alt="Screenshot 2025-01-24 at 12 57 47 AM" src="https://github.com/user-attachments/assets/6d3644ab-8515-4b23-9637-7328d4e19f63" />

**cc:** @amyeroberts @molbap @zucchini-nlp  kindly review this whenever you find the time .

sambhavnoobcoder committed:
change order of unmasking of tokens

sambhavnoobcoder committed:
library import

sambhavnoobcoder committed:
class setup

sambhavnoobcoder committed:
test function

sambhavnoobcoder committed:
refactor

sambhavnoobcoder committed:
add commit message

zucchini-nlp commented on tests/models/paligemma/test_attention_mask.py:
```diff
@@ -0,0 +1,62 @@
+import unittest
```
We need to add a test in `test_modeling_paligemma.py` file with tiny dummy model weights, and a single test covering attention with/without suffix should be enough

sambhavnoobcoder replied:
okay . I have done this in [96e1b43](https://github.com/huggingface/transformers/pull/35859/commits/96e1b43f4a04c88c08b26be1c75b53d5f8968ff1) commit , and removed the separate testing file from the PR . kindly review it for any more changes as well . if everything is fine , I'll push the fixed style changes and then we are good to merge .

zucchini-nlp reviewed (commented):
Thanks for opening a PR! Can you move the test to the `test_modeling_paligemma` and adapt it for dummy weights. I don't think we need slow test with big model, since we're just checking mask on attn weights

sambhavnoobcoder committed:
test modified

sambhavnoobcoder committed:
explict initiliasation of weights + made model smaller

sambhavnoobcoder committed:
removed sepete testing file

sambhavnoobcoder committed:
fixup

sambhavnoobcoder committed:
fixup core

zucchini-nlp reviewed (commented):
Thanks for iterating on this!

LGTM, but for the test we better just add a new one under `PaliGemmaForConditionalGenerationModelTest` which already creates dummy model and inputs. Sorry if it wasn't clear the first time
Also, let's test the model forward in general with `output_attentions=True`, instead of relying on a single `update_causal_mask` call. And add test if token type ids are passed vs when not passed, to make sure attention is masked correctly in all cases

sambhavnoobcoder committed:
test attention mask with token types

sambhavnoobcoder committed:
tests fixup

sambhavnoobcoder committed:
Merge branch 'main' into Paliegemma-Token-Mask

sambhavnoobcoder commented:
Thanks for the feedback! I've moved the test into `PaliGemmaForConditionalGenerationModelTest` class and am now using the existing dummy model setup instead of creating a separate test class. Instead of directly testing `_update_causal_mask` , I've switched to testing the full model forward pass with `output_attentions=True` which better reflects real usage. The test now covers both scenarios - with and without token_type_ids - to ensure attention masking works correctly in all cases. It verifies that pad tokens have zero attention weights across all attention layers and that the masking behavior remains consistent regardless of token type presence. Let me know if you'd like me to make any additional adjustments!

P.s also made the style fixup to make this ready to merge .

zucchini-nlp reviewed (approved):
Perfect, thanks a lot! Btw, you can remove the `PaliGemmaAttentionMaskTest` as it is testing the same thing

Requesting core maintainer's review and we'll merge

sambhavnoobcoder committed:
removed PaliGemmaAttentionMaskTest class

sambhavnoobcoder commented:
Thank you for the review @zucchini-nlp . I have removed `PaliGemmaAttentionMaskTest` in [a015a90](https://github.com/huggingface/transformers/pull/35859/commits/a015a90b12f5d3d524bbef20ba10cb2644edfc51) commit .


**cc:** @ArthurZucker for final review / merging.

sambhavnoobcoder committed:
Merge branch 'main' into Paliegemma-Token-Mask

molbap commented:
Thanks a lot! gentle ping @ArthurZucker for this! 
I'm sorry for not looking earlier, I thought we already moved that method around within `_update_causal_mask`. It turns out we are calling this method once more in `prepare_inputs_for_generation` when it's a PaliGemma2 model with HybridCache, any possible interaction with the training?

sambhavnoobcoder commented:
Thank you for the careful review @molbap  ! I've double-checked the interaction between `_update_causal_mask` and its usage in `prepare_inputs_for_generation` with HybridCache.

The current implementation maintains correct masking behavior in both contexts because:

1. The reordered masking operations (first prefix unmasking, then padding masks) ensure pad tokens stay masked regardless of where the method is called from
2. The `is_training` flag is properly propagated in both cases
3. Our test coverage in `test_attention_mask_with_token_types` verifies the masking behavior works correctly with and without token_type_ids

So the masking logic remains consistent whether called during regular training/inference or during generation with HybridCache, so no additional changes should be needed.

ArthurZucker reviewed (approved):
Cool! Thanks for fixing 🤗
